lib/node_modules/@stdlib/repl/lib/completer_engine.js

```diff
@@ -0,0 +1,747 @@
+	this._remainingLine = '';
+	this._completionPrefix = '';
+	this._completionsList = [];
+	this._highlightedCompletions = [];
+	this._completionsLength = [];
+	this._widthOfColumn = -1;
+});
+
+/**
+* Callback which should be invoked **before** a "keypress" event is processed by a readline interface.
+*
+* @name beforeKeypress
+* @memberof CompleterEngine.prototype
+* @type {Function}
+* @param {string} data - input data
+* @param {Object} key - key object
+* @returns {void}
+*/
+setNonEnumerableReadOnly( CompleterEngine.prototype, 'beforeKeypress', function beforeKeypress( data, key ) {
+	var cursor;
+	var line;
+
+	if ( !key ) {
+		this._ttyWrite.call( this._rli, data, key );
+		return;
+	}
+
+	// If user is already viewing completions, allow navigating it...
+	if ( this._isNavigating ) {
+		switch ( key.name ) {
+		// If user presses TAB while navigating, close the completer:
+		case 'tab':
+			debug( 'Received a TAB keypress event. Closing the completer engine...' );
+			this.closeCompleter();
+			break;
+
+		// If arrow keys detected, allow navigating the completions...
+		case 'down':
+			debug( 'Received a DOWN keypress event...' );
+			this._navigateDown();
+			break;
+		case 'up':
+			debug( 'Received an UP keypress event...' );
+			this._navigateUp( data, key );
+			break;
+		case 'left':
+			debug( 'Received a LEFT keypress event...' );
+			this._navigateLeft( data, key );
+			break;
+		case 'right':
+			debug( 'Received a RIGHT keypress event...' );
+			this._navigateRight( data, key );
+			break;
+		default:
+			this._ttyWrite.call( this._rli, data, key );
+		}
+		return;
+	}
+	// Trigger TAB completions:
+	cursor = this._rli.cursor;
+	line = this._rli.line;
+
+	// Get the line before the cursor:
+	this._inputLine = line.slice( 0, cursor );
+
+	// Get the line after the cursor:
+	this._remainingLine = line.slice( cursor );
+
+	// Pause the input stream before generating completions as the completer may be asynchronous...
+	this._rli.pause();
+	this._completer( this._inputLine, this._onCompletions );
+});
+
+/**
+* Callback for handling a "keypress" event.
+*
+* @name onKeypress
+* @memberof CompleterEngine.prototype
+* @type {Function}
+* @param {string} data - input data
+* @param {(Object|void)} key - key object
+* @returns {void}
+*/
+setNonEnumerableReadOnly( CompleterEngine.prototype, 'onKeypress', function onKeypress() {
+	var cursor;
+	var line;
+
+	// If user is not navigating completions, don't sustain completions...
+	if ( !this._isNavigating ) {
+		return;
+	}
+	// If user is trying to navigate, don't update completions based on the updated line...
+	if ( this._idx !== -1 ) {
+		return;
+	}
+	// Re-trigger completions based on updated line:
+	cursor = this._rli.cursor;
+	line = this._rli.line;
+
+	// If line is unchanged, no need to re-render completions panel..
+	if ( line === this._inputLine + this._remainingLine ) {
+		return;
+	}
+	// Clear invalid cache:
+	this.closeCompleter();
+
+	// Sustain the completer:
```

Comment:
By sustain, you mean ensuring the displayed completions remain visible?

Reply:
Yes.